src/libraries/System.Linq/src/System/Linq/Take.cs

```diff
@@ -20,6 +20,168 @@ public static IEnumerable<TSource> Take<TSource>(this IEnumerable<TSource> sourc
+                {
+                    return Empty<TSource>();
+                }
+            }
+            else if (!isEndIndexFromEnd)
+            {
+                return startIndex >= endIndex
+                    ? Empty<TSource>()
+                    : source.Skip(startIndex).Take(endIndex - startIndex);
+            }
+
+            return TakeIterator(source, isStartIndexFromEnd, startIndex, isEndIndexFromEnd, endIndex);
+        }
+
+        private static IEnumerable<TSource> TakeIterator<TSource>(
+            IEnumerable<TSource> source, bool isStartIndexFromEnd, int startIndex, bool isEndIndexFromEnd, int endIndex)
+        {
+            Debug.Assert(source != null);
+            Debug.Assert(isStartIndexFromEnd
+                ? startIndex > 0 && (!isEndIndexFromEnd || startIndex > endIndex)
+                : startIndex >= 0 && (isEndIndexFromEnd || startIndex < endIndex));
+            Debug.Assert(endIndex >= 0);
+
+            using IEnumerator<TSource> e = source.GetEnumerator();
+            if (isStartIndexFromEnd)
+            {
+                if (!e.MoveNext())
+                {
+                    yield break;
+                }
+
+                int index = 0;
+                Queue<TSource> queue = new();
+                queue.Enqueue(e.Current);
+
+                while (e.MoveNext())
+                {
+                    checked
+                    {
+                        index++;
+                    }
+
+                    if (queue.Count == startIndex)
+                    {
+                        queue.Dequeue();
+                    }
+
+                    queue.Enqueue(e.Current);
+                }
+
+                int count = checked(index + 1);
+                Debug.Assert(queue.Count == Math.Min(count, startIndex));
+
+                startIndex = count - startIndex;
+                if (startIndex < 0)
+                {
+                    startIndex = 0;
+                }
+
+                if (isEndIndexFromEnd)
+                {
+                    endIndex = count - endIndex;
+                }
+                else if (endIndex > count)
+                {
+                    endIndex = count;
+                }
+
+                Debug.Assert(endIndex - startIndex <= queue.Count);
+                for (int rangeIndex = startIndex; rangeIndex < endIndex; rangeIndex++)
+                {
+                    yield return queue.Dequeue();
+                }
+            }
+            else
+            {
+                if (!e.MoveNext())
+                {
+                    yield break;
+                }
+
+                int index = 0;
+                while (index < startIndex)
+                {
+                    if (!e.MoveNext())
+                    {
+                        yield break;
+                    }
+
+                    checked
+                    {
+                        index++;
+                    }
+                }
```

Comment:
I don't understand why the if block above is needed.  The goal here is to position the enumerator at the startIndex location, right?  If so, can't this just be:
```C#
int index = 0;
while (index < startIndex)
{
    if (!e.MoveNext())
    {
        yield break;
    }

    checked { index++; }
}
```
?

Reply:
Updated.